### ISSUE TITLE
Error: "/gcl-scripts/" could not be found on container with docker-type runner

### ISSUE THREAD

**Issue by ben-alkov** [firecow/gitlab-ci-local#298]
docker-type runner errors out with
```sh
gitlab-ci-local
parsing and downloads finished in 57 ms
uname starting registry.fedoraproject.org/fedora:34 (do_stuff)
uname pulled registry.fedoraproject.org/fedora:34 in 231 ms
uname copied to container in 365 ms
Error: "/gcl-scripts/" could not be found on container 4615c86204aa0eaecb04a4746a644ff1d6eca4c4255b18650068370949f87079: No suchfile or directory
```

**Minimal .gitlab-ci.yml illustrating the issue**
```yml
---
uname:
  image: registry.fedoraproject.org/fedora:34
  script:
    - uname -a
```

**Expected behavior**
No *public* url available.

**Host information**
Fedora 34
gitlab-ci-local 4.26.0

**Additional context**
I can't get e.g. `gitlab-runner exec docker uname` to work, as 'gitlab-runner' *requires* (seems to?) a "served" git repo to pull from (i.e. it ignores CWD, saying "fatal: 'FOO' does not appear to be a git repository. fatal: Could not read from remote repository").

Providing the actual upstream repo URL (using `--clone-url`) doesn't work due to cert issues (and neither `--docker-tlsverify=no` nor `variables: GIT_SSL_NO_VERIFY: "true"` in '.gitlabc-ci.yml' make any difference), which I'm pretty sure I have no hope of figuring out... which is why I'm here :)

**Comment by firecow** [comment]
```yml
---
uname:
  image: debian:stable-slim
  script:
    - uname -a
```
@ben-alkov 
Put this in a `.gitlab.yml` and run `gitlab-ci-local`
Paste the output here.

**Comment by firecow** [comment]
@ben-alkov I'm closing this for now..

Feel free to reopen, when you got time to run my `.gitlab-ci.yml` above

**Comment by ben-alkov** [comment]
Part of this is a combination of correctly configuring podman's docker socket "emulation", part of it is SELinux perms on the repo directories under test. With both of those things configured correctly, I am able to get `gitlab-runner exec docker uname` to work on the local repo as expected (frickin' sweet!).

I haven't had a chance to try with a different image, and given the above, I haven't re-tested with gitlab-ci-local

**Comment by firecow** [comment]
I have no idea why `gitlab-runner exec docker uname` is being involved :smile: 
In fact I have no idea what your are trying to do.
Which is why I want you run the simple example above, and paste the output...

**Comment by ben-alkov** [comment]
@firecow; Sorry for the confusion.

Last time, I was able to do what I needed using only the actual `gitlab-runner`. *This* time, I need to actually get `gitlab-ci-local` to work, so...

Using _exactly_ the '.gitlab-ci.yml' you requested:


```
$ gitlab-ci-local
parsing and downloads finished in 33 ms
uname starting debian:stable-slim (test)
uname pulled debian:stable-slim in 3.09 s
uname copied to container in 336 ms
Error: Command failed with exit code 125: docker cp .gitlab-ci-local/scripts/. 6b1915bf6dbb8eedb682312e60bc1b13bd18cad258f1a5ee6be63908ddb1e661:/gcl-scripts/
Error: "/gcl-scripts/" could not be found on container 6b1915bf6dbb8eedb682312e60bc1b13bd18cad258f1a5ee6be63908ddb1e661: No such fileor directory
    at makeError (/snapshot/firecow-gitlab-ci-local/node_modules/execa/lib/error.js:60:11)
    at handlePromise (/snapshot/firecow-gitlab-ci-local/node_modules/execa/index.js:118:26)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at Job.execScripts (/snapshot/firecow-gitlab-ci-local/src/job.ts:633:13)
    at Job.start (/snapshot/firecow-gitlab-ci-local/src/job.ts:363:36)
```

P.S. This is with the very latest Linux release from here (4.31.0), running now on Fedora 35

**Comment by firecow** [comment]
What says `docker --version`

**Comment by ben-alkov** [comment]
Well, it's `podman`; version 3.4.7

**Comment by firecow** [comment]
Please try with docker...

**Comment by ben-alkov** [comment]
No docker here.

**Comment by firecow** [comment]
Well, its called a docker-executor 🤣

Or you could make a feature request for supporting custom-executors in gitlab-ci-local.

**Comment by ben-alkov** [comment]
Sorry, I was being a teensy bit flippant. The "official" docker package isn't available for Fedora without jumping through extraordinary and annoying hoops; thus there's only podman here.

OTOH, the 'docker' command is aliased to 'podman'. It's a near-exact replacement for docker, without replicating any of docker's bugs or lax security. If either of those things (bugs or lax security) comes into play during `gitlab-ci-local` execution, then it could be an issue.

**Comment by ben-alkov** [comment]
I'm not an expert on any of these things, but my gut says that maybe there's an issue trying to mount/copy/something else the contents of the newly created '.gitlab-ci-local/scripts/' directory into the pulled container...

**Comment by ben-alkov** [comment]
> Or you could try to alias all docker commands to point to podman.

Just to be clear: as I mentioned above, this _is_ the case here

**Comment by firecow** [comment]
We are not going to replace docker with podman until gitlab does.

So I'm affraid you are on your own.

But docker cp certainly isn't hitting its podman equivilant...

**Comment by ben-alkov** [comment]
> But docker cp certainly isn't hitting its podman equivilant...

Can you explain what it's trying to do? ` docker cp .gitlab-ci-local/scripts/. 6b1915bf6dbb8eedb682312e60bc1b13bd18cad258f1a5ee6be63908ddb1e661:/gcl-scripts/` seems quite straightforward, save the fact that the '/gcl-scripts/' directory *definitely* doesn't already exist in 'debian:stable-slim'...

**Comment by firecow** [comment]
Yes, docker cp is quite straightforward.

Error: Command failed with exit code 125: docker cp

And your "alias" or lack thereof doesn't hit pod or podman doesn't understand the arguments or needs more arguments or the arguments aren't in the right order.

**Comment by ben-alkov** [comment]
`docker cp [OPTIONS] SRC_PATH CONTAINER:DEST_PATH` practically verbatim from https://docs.docker.com/engine/reference/commandline/cp/

`podman cp [options] src_path [container:]dest_path` practically verbatim from https://docs.podman.io/en/latest/markdown/podman-cp.1.html

so I don't understand your assertion that the issue is somehow with podman.

The traceback literally says `"Error: "/gcl-scripts/" could not be found on container 6b1915bf6dbb8eedb682312e60bc1b13bd18cad258f1a5ee6be63908ddb1e661: No such fileor directory"`, so again... I'm wondering where that directory is supposed to come from; I seriously doubt that it's a part of the standard 'debian:stable-slim' image.

Is the container runtime supposed to create that directory in the container?

**Comment by firecow** [comment]
`Error: Command failed with exit code 125: docker cp`

That is the important error...

gitlab-ci-local calls docker cp, not podman cp.

So again, what does your docker --version say?

**Comment by ben-alkov** [comment]
This is `docker` on a podman system:

```sh
#!/usr/bin/sh
[ -f /etc/containers/nodocker ] || \
echo "Emulate Docker CLI using podman. Create /etc/containers/nodocker to quiet msg." >&2
exec /usr/bin/podman "$@"
```

**Comment by ben-alkov** [comment]
```
docker --version
podman version 3.4.7
```

**Comment by firecow** [comment]
docker cp must be different from podman cp in some way the.

**Comment by ben-alkov** [comment]
> docker cp must be different from podman cp in some way then

Apparently.

It would help me further troubleshoot if I understood what it's trying to do.

Is the '/gcl-scripts/' directory somehow being created prior to the `docker cp` command, or is it automatically created *by* the `docker cp` command?

**Comment by firecow** [comment]
You can see that in job.ts

But I'm pretty sure its just a copy from .gitlab-ci-local/scripts onto the containers gcl-scripts

**Comment by ben-alkov** [comment]
I wish I could read Typescript, but I'm not really fluent.

> the containers /gcl-scripts

I'm still not clear how this comes into existence. Is it already in the debian image? Is there some silent step earlier in the process which creates it in the local container?

**Comment by ben-alkov** [comment]
On further inspection of the Typescript source and my local system, it looks like the container $FOO in `docker cp .gitlab-ci-local/scripts/. $FOO:/gcl-scripts/` is **NOT** debian, but rather 'docker.io/firecow/gitlab-ci-local-util', so I guess that dir already exists in the image.

**Comment by firecow** [comment]
```
mjn@mjn-laptop:~$ docker run -it firecow/gitlab-ci-local-util bash
root@7fc5aae56499:/# ls
bin  boot  dev	etc  home  lib	lib64  media  mnt  opt	proc  root  run  sbin  srv  sys  tmp  usr  var
root@7fc5aae56499:/# 
```
Nope, guess again...

The word `gcl-scripts` is mentioned two places in the code.
```
if (this.imageName) {
    await Utils.spawn(["docker", "cp", `${stateDir}/scripts/.`, `${this._containerId}:/gcl-scripts/`], cwd);
}
```

and here

```
if (this.imageName) {
    cp.stdin?.end(`/gcl-scripts/${safeJobName}`);
} else {
    cp.stdin?.end(`./${stateDir}/scripts/${safeJobName}`);
}
```

**Comment by firecow** [comment]
https://github.com/firecow/gitlab-ci-local/pull/492/commits/e2ac3cd86e7c4054b1ee16b46ef7c7ddbb5669d8 I'll try to remedy podman's quirk by putting the gcl command script in `/`

**Comment by ben-alkov** [comment]
:tada: tested locally; fixed in #492! @firecow TYVM sir!
